go.mod

```diff
@@ -57,3 +57,99 @@ require (
 	google.golang.org/grpc v1.32.0
 	google.golang.org/protobuf v1.24.0
 )
+
+require (
```

Comment:
why is this split into a second require block?

Reply:
I really don't know for sure. That file is managed by go directly.

Reply:
this is something due to golang 1.17 and changes in the way go.mod is processed

Reply:
It is because of go 1.17!

`Lazy modules can still be consumed by pre-1.17 users, and will have the same or higher selected versions of their dependencies when consumed by those users.
Since all of the modules in the wild today are eager, the main difference you'll notice today is that your go.mod file will contain more entries (so that it explicitly covers all of the packages transitively imported by your module). As your dependencies upgrade to become lazy, you should also notice your go.sum files shrink (sometimes dramatically), and you may find that it is easier to upgrade away from unwanted transitive dependencies.`

https://khanakia.medium.com/go-1-17-compiler-got-faster-and-go-mod-api-changed-71f35267d616

Reply:
I think if you remove lines 59-61 you'll be able to combine all the entries into one require block. There are indirect module deps in both require blocks.

Reply:
I can do that, but golang will process the file again when executing `go mod tidy` and leave it the same again.

Reply:
i tried that just now and that's not true. `go mod tidy` does not split/merge require blocks, so if you combine into 1 require block it will sort the whole list, and leave it as the more typical 1 require block.

Reply:
I think we will encounter this many times in the future as things get upgraded to >=1.17
We can revert the automatic change.. but I have the feeling that this split will be the typical format in future implementations. 

This is all explained in the 1.17 release notes: https://golang.org/doc/go1.17#go-command

In particular:

> Because the number of explicit requirements may be substantially larger in an expanded Go 1.17 go.mod file, the newly-added requirements on indirect dependencies in a go 1.17 module are maintained in a separate require block from the block containing direct dependencies.
>
> To facilitate the upgrade to Go 1.17 pruned module graphs, the go mod tidy subcommand now supports a -go flag to set or change the go version in the go.mod file. To convert the go.mod file for an existing module to Go 1.17 without changing the selected versions of its dependencies, run:
>
>  go mod tidy -go=1.17
> By default, go mod tidy verifies that the selected versions of dependencies relevant to the main module are the same versions that would be used by the prior Go release (Go 1.16 for a module that specifies go 1.17), and preserves the go.sum entries needed by that release even for dependencies that are not normally needed by other commands.
>
> The -compat flag allows that version to be overridden to support older (or only newer) versions, up to the version specified by the go directive in the go.mod file. To tidy a go 1.17 module for Go 1.17 only, without saving checksums for (or checking for consistency with) Go 1.16:
>
>  go mod tidy -compat=1.17

Reply:
alright, it does look like this is the direction the canonical module file will look in the future :+1: